lib/datadog/remote-config/src/lib.rs

```diff
@@ -0,0 +1,481 @@
+use std::collections::{HashMap, HashSet};
+
+use futures::AsyncReadExt;
+use hyper::{client::HttpConnector, Body, Method};
+use hyper_openssl::HttpsConnector;
+use prost::Message;
+use rand::distributions::{Alphanumeric, DistString};
+use serde::{Deserialize, Serialize};
+use tuf::{
+    crypto::HashAlgorithm,
+    interchange,
+    metadata::{Metadata, MetadataPath, MetadataVersion, RawSignedMetadata},
+    repository::{EphemeralRepository, RepositoryProvider, RepositoryStorage},
+};
+
+mod proto {
+    include!(concat!(env!("OUT_DIR"), "/datadog.config.rs"));
+}
+mod error;
+
+use error::Result;
+use proto::{DelegatedMeta, LatestConfigsRequest, LatestConfigsResponse, TopMeta};
+
+pub use error::Error;
+pub use tuf::metadata::{TargetDescription, TargetPath};
+
+type TUFClient = tuf::client::Client<
+    interchange::Json,
+    EphemeralRepository<interchange::Json>,
+    EphemeralRepository<interchange::Json>,
+>;
+
+// TODO: are these safe to commit/embed?
+const CONFIG_ROOT: &[u8] = include_bytes!("../config_root.json");
+const DIRECTOR_ROOT: &[u8] = include_bytes!("../director_root.json");
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Config {
+    pub site: String,
+    pub api_key: String,
+    pub app_key: String,
+    pub hostname: String,
+    pub agent_version: String,
+}
+
+#[derive(Debug)]
+struct Inner {
+    http: hyper::Client<HttpsConnector<HttpConnector>>,
+    site: String,
+    api_key: String,
+    app_key: String,
+}
+
+pub struct Client {
+    inner: Inner,
+    hostname: String,
+    agent_version: String,
+    products: HashSet<String>,
+    new_products: HashSet<String>,
+    active_client: proto::Client,
+    backend_client_state: Vec<u8>,
+    config_client: TUFClient,
+    director_client: TUFClient,
+}
+
+impl Client {
+    pub async fn initialize(config: Config) -> Result<Self> {
+        let Config {
+            site,
+            api_key,
+            app_key,
+            hostname,
```

Comment:
Curious around the requirements for this, and how the hostname will be provided/obtained.